vm/vm_jit.go

```diff
@@ -1,31 +1,376 @@
+		b := big.NewInt(int64(m[i]))
+		b.Lsh(b, uint(i)*8)
+		n.Add(n, b)
+	}
+	return n
+}
+
+func llvm2bytes(data *byte, length uint64) []byte {
+	if length == 0 {
+		return nil
+	}
+	if data == nil {
+		panic("llvm2bytes: nil pointer to data")
+	}
+	return (*[1 << 30]byte)(unsafe.Pointer(data))[:length:length]
+}
+
+func untested(condition bool, message string) {
+	if condition {
+		panic("Condition `" + message + "` tested. Remove assert.")
+	}
+}
+
+func assert(condition bool, message string) {
+	if !condition {
+		panic("Assert `" + message + "` failed!")
+	}
 }
 
 func NewJitVm(env Environment) *JitVm {
-	backupVm := New(env)
-	return &JitVm{env: env, backup: backupVm}
+	return &JitVm{env: env}
 }
 
 func (self *JitVm) Run(me, caller ContextRef, code []byte, value, gas, price *big.Int, callData []byte) (ret []byte, err error) {
-	return self.backup.Run(me, caller, code, value, gas, price, callData)
+	self.env.SetDepth(self.env.Depth() + 1)
+
+	if Precompiled[string(me.Address())] != nil {
+		// if it's address of precopiled contract
+		// fallback to standard VM
+		stdVm := New(self.env)
+		return stdVm.Run(me, caller, code, value, gas, price, callData)
+	}
+
+	self.me = me // FIXME: Make sure Run() is not used more than once
+	self.callerAddr = caller.Address()
+	self.price = price
+
+	self.data.elems[Gas] = big2llvm(gas)
+	self.data.elems[address] = address2llvm(self.me.Address())
+	self.data.elems[Caller] = address2llvm(caller.Address())
+	self.data.elems[Origin] = address2llvm(self.env.Origin())
+	self.data.elems[CallValue] = big2llvm(value)
+	self.data.elems[CallDataSize] = big2llvm(big.NewInt(int64(len(callData)))) // TODO: Keep call data size as i64
+	self.data.elems[GasPrice] = big2llvm(price)
+	self.data.elems[CoinBase] = address2llvm(self.env.Coinbase())
+	self.data.elems[TimeStamp] = big2llvm(big.NewInt(self.env.Time())) // TODO: Keep timestamp as i64
+	self.data.elems[Number] = big2llvm(self.env.BlockNumber())
+	self.data.elems[Difficulty] = big2llvm(self.env.Difficulty())
+	self.data.elems[GasLimit] = big2llvm(self.env.GasLimit())
+	self.data.elems[CodeSize] = big2llvm(big.NewInt(int64(len(code)))) // TODO: Keep code size as i64
+	self.data.callData = getDataPtr(callData)
+	self.data.code = getDataPtr(code)
+
+	result := C.evmjit_run(unsafe.Pointer(&self.data), unsafe.Pointer(self))
+	//fmt.Printf("JIT result: %d\n", r)
+
+	if result.returnCode >= 100 {
+		err = errors.New("OOG from JIT")
+		gas.SetInt64(0) // Set gas to 0, JIT does not bother
+	} else {
+		gasLeft := llvm2big(&self.data.elems[Gas]) // TODO: Set value directly to gas instance
+		gas.Set(gasLeft)
+		if result.returnCode == 1 { // RETURN
+			ret = C.GoBytes(result.returnData, C.int(result.returnDataSize))
+			C.free(result.returnData)
+		} else if result.returnCode == 2 { // SUICIDE
+			state := self.Env().State()
+			receiverAddr := llvm2hash(bswap(&self.data.elems[address]))
+			receiverAddr = trim(receiverAddr) // TODO: trim all zeros or subslice 160bits?
```

Comment:
Availability: Until the GC thinks it can be freed. 

It's always 20 bytes https://github.com/ethereum/go-ethereum/blob/develop/ethutil/bytes.go#L214